### ISSUE TITLE
tgt: iscsiadm failed to logout (timeout) in some situation

### ISSUE THREAD

**Issue by yasker** [rancher/longhorn#230]
As drone shows:
https://drone.rancher.io/rancher/longhorn/50
```
127.0.0.1 - - [01/Nov/2016:23:01:28 +0000] "DELETE /v1/replicas/1 HTTP/1.1" 200 268
time="2016-11-01T23:01:28Z" level=info msg="Shutdown SCSI device for 172.17.0.6:iqn.2014-07.com.rancher:test-volume" 
time="2016-11-01T23:01:28Z" level=error msg="Replicator.ReadAt:1 Volume no longer exist" 
time="2016-11-01T23:02:28Z" level=error msg="Setting replica tcp://localhost:9505 to ERR due to: Volume no longer exist" 
time="2016-11-01T23:02:28Z" level=error msg="Error when shutting down frontend:Fail to stop SCSI device: Fail to logout target: Timeout executing: nsenter [--mount=/host/proc/1/ns/mnt --net=/host/proc/1/ns/net iscsiadm -m node -T iqn.2014-07.com.rancher:test-volume --logout -p 172.17.0.6], output , error <nil>" 
time="2016-11-01T23:02:28Z" level=error msg="Ignoring error because tcp://localhost:9502 is mode RW: tcp://localhost:9505: Volume no longer exist" 
time="2016-11-01T23:02:28Z" level=info msg="Closing: localhost:9502" 
```

Seems if backend doesn't response when logging out can result in this kind of scenario. The sequence above seems show that read/write happens before backend was removed from controller, thus caused trouble for logout.

**Comment by yasker** [comment]
log from iscsid in syslog
```
Nov  7 19:41:56 drone-agent0 iscsid: Connection116:0 to [target: iqn.2014-09.com.rancher:test-volume, portal: 172.17.0.14,3260] through [iface: default] is shutdown.
Nov  7 19:41:56 drone-agent0 iscsid: Connection117:0 to [target: iqn.2014-09.com.rancher:test-volume, portal: 172.17.0.14,3260] through [iface: default] is operational now

Nov  7 19:42:05 drone-agent0 kernel: [929880.086779] unregister_netdevice: waiting for veth1444114 to become free. Usage count = 1
Nov  7 19:42:05 drone-agent0 kernel: [929880.454915] unregister_netdevice: waiting for vethb5fba63 to become free. Usage count = 1
Nov  7 19:42:14 drone-agent0 dd.collector[2006]: CRITICAL (docker_daemon.py:213): Error while fetching server API version: ('Connection aborted.', error(13, 'Permission denied'))
Nov  7 19:42:14 drone-agent0 dd.collector[2006]: WARNING (__init__.py:679): Initialization failed. Will retry at next iteration
Nov  7 19:42:15 drone-agent0 kernel: [929890.234753] unregister_netdevice: waiting for veth1444114 to become free. Usage count = 1
Nov  7 19:42:15 drone-agent0 kernel: [929890.574890] unregister_netdevice: waiting for vethb5fba63 to become free. Usage count = 1
Nov  7 19:42:25 drone-agent0 kernel: [929900.494786] unregister_netdevice: waiting for veth1444114 to become free. Usage count = 1
Nov  7 19:42:25 drone-agent0 kernel: [929900.846922] unregister_netdevice: waiting for vethb5fba63 to become free. Usage count = 1
Nov  7 19:42:26 drone-agent0 multipathd: sdc: remove path (uevent)
Nov  7 19:42:28 drone-agent0 dd.forwarder[2005]: INFO (transaction.py:167): Flushing 1 transaction during flush #305680
Nov  7 19:42:33 drone-agent0 dd.collector[2006]: CRITICAL (docker_daemon.py:213): Error while fetching server API version: ('Connection aborted.', error(13, 'Permission denied'))
Nov  7 19:42:33 drone-agent0 dd.collector[2006]: WARNING (__init__.py:679): Initialization failed. Will retry at next iteration
Nov  7 19:42:35 drone-agent0 kernel: [929910.846850] unregister_netdevice: waiting for veth1444114 to become free. Usage count = 1
Nov  7 19:42:35 drone-agent0 kernel: [929910.966906] unregister_netdevice: waiting for vethb5fba63 to become free. Usage count = 1
Nov  7 19:42:46 drone-agent0 kernel: [929921.102886] unregister_netdevice: waiting for vethb5fba63 to become free. Usage count = 1
Nov  7 19:42:46 drone-agent0 kernel: [929921.186917] unregister_netdevice: waiting for veth1444114 to become free. Usage count = 1
Nov  7 19:42:53 drone-agent0 dd.collector[2006]: CRITICAL (docker_daemon.py:213): Error while fetching server API version: ('Connection aborted.', error(13, 'Permission denied'))
Nov  7 19:42:53 drone-agent0 dd.collector[2006]: WARNING (__init__.py:679): Initialization failed. Will retry at next iteration
Nov  7 19:42:56 drone-agent0 kernel: [929931.258881] unregister_netdevice: waiting for veth1444114 to become free. Usage count = 1
Nov  7 19:42:56 drone-agent0 kernel: [929931.346909] unregister_netdevice: waiting for vethb5fba63 to become free. Usage count = 1
Nov  7 19:42:56 drone-agent0 iscsid: session in invalid state for logout. Try again later
Nov  7 19:42:57 drone-agent0 iscsid: Connection117:0 to [target: iqn.2014-09.com.rancher:test-volume, portal: 172.17.0.14,3260] through [iface: default] is shutdown.
Nov  7 19:42:57 drone-agent0 iscsid: IPC qtask write failed: Broken pipe
```

Nothing showing for `/var/log/tgtd.log` inside container.

**Comment by yasker** [comment]
I cannot really verify the fix of this issue since it's hard to reproduce. Feel free to reopen in the future if it happenes again.

**Comment by yasker** [comment]
No, it's still happening... Reopen the bug.

**Comment by yasker** [comment]
Reproduced with Ubuntu 14.04.5 LTS and SMP = 4, memory = 4G on KVM virtual machine.

**Comment by yasker** [comment]
logs in dmesg:

```
[  167.307970] sd 2:0:0:1: [sda] tag#0 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[  167.310036] sd 2:0:0:1: [sda] tag#0 Sense Key : Medium Error [current] 
[  167.311544] sd 2:0:0:1: [sda] tag#0 Add. Sense: Unrecovered read error
[  167.313250] sd 2:0:0:1: [sda] tag#0 CDB: Read(10) 28 00 00 00 00 08 00 00 08 00
[  167.314939] blk_update_request: critical medium error, dev sda, sector 8
[  167.316599] Buffer I/O error on dev sda, logical block 1, async page read
[  167.318263] Dev sda: unable to read RDB block 8
```

**Comment by yasker** [comment]
At least, in this case, the issue was triggered by immediate shutdown ISCSI client after brought it up. I've tried to add 5 seconds interval before the logout operation, then the issue disappears.

Also, it takes about 80 seconds for iscsid responses to the logout.

**Comment by yasker** [comment]
It's because of current TGT Longhorn code doesn't have a mechanism for the server (longhorn controller) to notify client (TGT) of disconnection. Also after disconnection, the existing commands in the queue weren't drained.

I've added two fixes for it:
1. One more message type was added: TypeClose. It's used to notify close of disconnection.
2. The timeout handler was added to the command handler. It's 15 seconds by default, which is the same of default in Longhorn controller/replica communication.
